### DISCUSSION TITLE
Updating config during pre-update-cmd

### DISCUSSION THREAD

**Discussion by voidgraphics** [composer/composer#10624] [Q&A / Support] [answered]
I'm having trouble overriding the repositories before running composer update.

I have a `composer.local.json` config file that contains the proper repository definitions to allow local development via symlinks.
I have created a PHP class that runs on the `pre-update-cmd` event, which conditionally merges those repositories into the config, depending on the environment being local or production.

Something like:

```php
$local_composer_config = json_decode(file_get_contents($composer_local_json_path), true);

$event->getComposer()->getConfig()->merge($local_composer_config);
```
After that, if I inspect the config instance, I can see my repositories have been properly set, as expected:

![CleanShot 2022-03-16 at 14 07 35](https://user-images.githubusercontent.com/9298484/158596902-3d194efb-290e-4c5a-8340-574e5c689c3e.png)

Even if I inspect the config in the `post-update-cmd` event, the config does contain the right repositories, according to my changes. 

However, the symlinks are not generated. It is still installing from the vcs repositories that are defined in the base composer.json file.

Am I facing a bug? Or are we not supposed to be able to override the config in such a way?

Thanks!

**Comment by Seldaek** [comment, marked as answer]
I'd suggest looking at https://github.com/wikimedia/composer-merge-plugin and https://github.com/franzliedke/studio for two plugins that do help with merging & local dev.

Probably you can use either of those and drop your homegrown solution entirely.

**Reply by voidgraphics** [reply]
Thanks for the suggestion, composer-merge-plugin is close to what I need. It does have an annoyance though: since it's also installed via composer, it does not work when doing a fresh install, because it hasn't been installed yet.

At least that's what I think is happening, because when I delete the vendor directory and run `composer install` or `composer update` the first time, it does not do the symlinks, but on the second try, it finally does. From a quick glance, it seems studio will have the same drawback.

Forgive me for insisting, but is my first approach not supposed to work at all? The config class does have a merge method, which has special instructions specifically to merge repositories and everything, and it seems to be doing a good job at it ; so why is it not actually using the merged config when it does the install/update right after?

Thanks for your time!

**Reply by Seldaek** [reply]
Yeah those kind of plugins are really better installed globally using `composer global require foo/bar`, otherwise you indeed end up with issues on the first/fresh install.

> Forgive me for insisting, but is my first approach not supposed to work at all? The config class does have a merge method, which has special instructions specifically to merge repositories and everything, and it seems to be doing a good job at it ; so why is it not actually using the merged config when it does the install/update right after?

The problem with the way you did it is by the time pre-update-cmd runs it's too late already, repositories have already been instantiated. You'd have to also modify the RepositoryManager ($composer->getRepositoryManager()) and add/prepend the new repository you are adding to it. That might fix things.

**Reply by voidgraphics** [reply]
That works great! 

For anyone else who might want to do this in the future, a quick and dirty example of what you can do in the PHP event handler:

`composer.json`:
```json
{
    "scripts": {
        "pre-install-cmd": "Foo::loadLocalRepositories",
        "pre-update-cmd": "Foo::loadLocalRepositories",
    }
}
```


`Foo.php`
```php
public static function loadLocalRepositories(Event $event)
{
    // Ideally you would load this from a json file, like `composer.local.json` or something
    $local_repositories = [
        'foo' => [
            'type' => 'path',
            'url' => '../foo',
            'options' => ['symlink' => true]
        ]
    ];

    $repositoryManager = $event->getComposer()->getRepositoryManager();

    foreach ($local_repositories as $name => $config) {
        $repository = $repositoryManager->createRepository($config['type'], $config, (string) $name);

        $repositoryManager->prependRepository($repository);
    }
}
```

Thank you @Seldaek.
